packages/core/src/demo/echo.js

```diff
@@ -91,11 +91,53 @@ const messageEchoCache = {};
 
   account.on(PayloadBundleType.TEXT, async data => {
     const {content, id: messageId} = data;
+    let textPayload;
+
+    const linkPreviews = [];
+
+    if (content.linkPreview) {
+      for (const linkPreview of content.linkPreview) {
+        let linkPreviewImage;
+
+        if (linkPreview.article && linkPreview.article.image) {
+          const originalLinkPreviewImage = linkPreview.article.image;
+          const imageBuffer = await account.service.conversation.getAsset(originalLinkPreviewImage.uploaded);
+
+          linkPreviewImage = {
+            data: imageBuffer,
+            height: originalLinkPreviewImage.original.image.height,
+            type: originalLinkPreviewImage.original.mimeType,
+            width: originalLinkPreviewImage.original.image.width,
+          };
+        }
+
+        const newLinkPreview = await account.service.conversation.createLinkPreview(
+          linkPreview.url,
+          linkPreview.urlOffset,
+          linkPreview.permanentUrl,
+          linkPreviewImage,
+          linkPreview.summary,
+          linkPreview.title,
+          linkPreview.tweet
+        );
 
-    const textPayload = account.service.conversation.createText(content.text);
-    messageEchoCache[messageId] = textPayload.id;
+        linkPreviews.push(newLinkPreview);
+      }
 
-    await handleIncomingMessage(data);
+      await handleIncomingMessage(data);
+
+      const messageIdOriginal = messageIdCache[messageId];
+
+      if (messageIdOriginal) {
+        textPayload = account.service.conversation.createText(content.text, linkPreviews, messageIdOriginal);
+      } else {
+        logger.warn(`Link preview for message ID "${messageId} was received before the original message."`);
```

Comment:
I think this is a case we handle in the webapp. If the preview comes before the actual text message, we save it as is. (I believe there was some stories like this in the past with mobile client + bad networking or something).
// ping @gregor

Reply:
Ok nevermind, this is just demo code.